File: 2019/18/alt18.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"math"
)

type Point struct {
	x int
	y int
}

type State struct {
	pos  Point
	keys KeySet
}

type KeySet int32

func isKey(b byte) bool {
	return b >= 'a' && b <= 'z'
}

func isDoor(b byte) bool {
	return b >= 'A' && b <= 'Z'
}

func assert(b bool, s string) {
	if !b {
		panic(s)
	}
}

func (k KeySet) Add(b byte) KeySet {
	assert(isKey(b), "invalid key")
	return k | (1 << (b - 'a'))
}

func (k KeySet) Has(b byte) bool {
	assert(isKey(b), "invalid key")
	return (k & (1 << (b - 'a'))) != 0
}

func read_data(inp []byte) ([][]byte, map[byte]Point, int) {
	gr := [][]byte{}
	loc := make(map[byte]Point)
	nKeys := 0
	for y, row := range bytes.Split(inp, []byte("\n")) {
		if len(row) == 0 {
			break
		}
		gr = append(gr, row)
		if len(row) != len(gr[0]) {
			panic("Invalid input")
		}
		for x, c := range row {
			if c != '.' && c != '#' {
				loc[c] = Point{x, y}
				if isKey(c) {
					nKeys++
				}
			}
		}
	}
	return gr, loc, nKeys
}

func move(gr [][]byte, from Point, dr int) (Point, byte) {
	var to Point
	switch dr {
	case 0:
		to = Point{from.x, from.y - 1}
	case 1:
		to = Point{from.x + 1, from.y}
	case 2:
		to = Point{from.x, from.y + 1}
	case 3:
		to = Point{from.x - 1, from.y}
	default:
		panic("invalid direction")
	}
	if to.x < 0 || to.x >= len(gr[0]) || to.y < 0 || to.y >= len(gr) {
		return to, '#'
	}
	return to, gr[to.y][to.x]
}

func keyForDoor(c byte) byte {
	assert(isDoor(c), "invalid door")
	return c + 32
}

func keyDistances(gr [][]byte, st State) map[byte]int {
	dst := make(map[byte]int)
	vs := make([][]bool, len(gr))
	for i := 0; i < len(gr); i++ {
		vs[i] = make([]bool, len(gr[0]))
	}

	type PD struct {
		pos  Point
		dist int
	}
	var q = make([]PD, 1, 10000)
	q[0] = PD{st.pos, 0}

	for len(q) != 0 {
		var pd PD
		pd, q = q[0], q[1:]
		if !vs[pd.pos.y][pd.pos.x] {
			vs[pd.pos.y][pd.pos.x] = true
			c := gr[pd.pos.y][pd.pos.x]
			if isKey(c) && !st.keys.Has(c) {
				dst[c] = pd.dist
			}
			for dr := 0; dr < 4; dr++ {
				p, cc := move(gr, pd.pos, dr)
				if cc == '#' || (isDoor(cc) && !st.keys.Has(keyForDoor(cc))) {
					continue
				}
				q = append(q, PD{p, pd.dist + 1})
			}
		}
	}

	return dst
}

func part1(instr []byte) int {
	gr, loc, nKeys := read_data(instr)
	pathLengths := make(map[State]int)
	pathLengths[State{loc['@'], 0}] = 0
	for depth := 0; depth < nKeys; depth++ {
		fmt.Println("GEN", depth, "SIZE", len(pathLengths))
		oldPathLengths := pathLengths
		pathLengths = make(map[State]int)
		for state, totalDist := range oldPathLengths {
			kds := keyDistances(gr, state)
			for k, dist := range kds {
				newTotalDist := totalDist + dist
				newPos := loc[k]
				newKeys := state.keys.Add(k)
				newState := State{newPos, newKeys}
				if pl, ok := pathLengths[newState]; !ok || newTotalDist < pl {
					pathLengths[newState] = newTotalDist
				}
			}
		}
	}

	minDist := math.MaxInt64
	var minState State
	for state, totalDist := range pathLengths {
		if totalDist < minDist {
			minDist = totalDist
			minState = state
		}
	}

	fmt.Println(minState)
	return minDist
}

func runTests() {
	test2 := []byte(`########################
#f.D.E.e.C.b.A.@.a.B.c.#
######################.#
#d.....................#
########################`)
	fmt.Println("TEST2", part1(test2))

	test3 := []byte(`########################
#...............b.C.D.f#
#.######################
#.....@.a.B.c.d.A.e.F.g#
########################`)
	fmt.Println("TEST3", part1(test3))

	test4 := []byte(`#################
#i.G..c...e..H.p#
########.########
#j.A..b...f..D.o#
########@########
#k.E..a...g..B.n#
########.########
#l.F..d...h..C.m#
#################`)
	fmt.Println("TEST4", part1(test4))

	test5 := []byte(`########################
#@..............ac.GI.b#
###d#e#f################
###A#B#C################
###g#h#i################
########################`)
	fmt.Println("TEST5", part1(test5))
}

func main() {
	//defer profile.Start().Stop()
	runTests()

	inBytes, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic("input")
	}
	ans1 := part1(inBytes)
	fmt.Println("PART1", ans1)
}

File: 2019/18/alt18.js
'use strict';
var assert = require('assert');
var fs = require('fs');

const isKey = c => c >= 'a' && c <= 'z';
const isDoor = c => c >= 'A' && c <= 'Z';
function keyForDoor(c) {
    assert(isDoor(c), "Invalid door in keyForDoor");
    return c.toLowerCase();
}

class KeySet {
    constructor(n=0) {
        this.n = n;
    }
    keyNum(b) {
        assert(isKey(b), "invalid key");
        return b.charCodeAt(0) - 97;
    }
    add(b) {
        let k = this.keyNum(b);
        this.n += 1 << k;
        return this;
    }
    has(b) {
        let k = this.keyNum(b);
        return this.n & (1 << k);
    }
}


function readData(inStr) {
    let gr = [];
    let loc = {};
    let nKeys = 0;
    let lines = inStr.split("\n");
    for (const [y,row1] of lines.entries()) {
        const row = row1.trim();
        if (row == "") continue;
        gr.push(row);
        assert(row.length == gr[0].length);
        for (const [x,c] of row.split("").entries()) {
            if (c != '.' && c != '#') {
                loc[c] = [x, y];
                if (isKey(c)) nKeys++;
            }
        }
    }
    return [gr, loc, nKeys];
}

const gridSize = (gr) => {return {nr: gr.length, nc: gr[0].length}};

function move(gr, x, y, dr) {
    let toX = 0;
    let toY = 0;
    switch (dr) {
        case 0:
            toX = x;
            toY = y-1;
            break;
        case 1:
            toX = x + 1;
            toY = y;
            break;
        case 2:
            toX = x;
            toY = y+1;
            break;
        case 3:
            toX = x - 1;
            toY = y;
            break;
        default:
            assert(false, "invalid direction");
    }
    let {nr, nc} = gridSize(gr);
    if (toX < 0 || toX >= nc || toY < 0 || toY >= nr) {
        return [toX, toY, '#'];
    }
    //return [toX, toY, gr[toY][toX]];
    return {x: toX, y: toY, c: gr[toY][toX]};
}

function keyDistances(gr, fromX, fromY, keys) {
    let dst = {};
    let {nr, nc} = gridSize(gr);
    let vs = new Array(nr);
    for (let y = 0; y < nr; y++) {
        vs[y] = new Array(gr.length).fill(false);
    }

    let q = [{x: fromX, y: fromY, dist: 0}];
    while (q.length != 0) {
        let {x, y, dist} = q.shift();
        if (!vs[y][x]) {
            vs[y][x] = true;
            let c = gr[y][x];
            if (isKey(c) && !keys.has(c)) {
                dst[c] = dist;
            }
            for (let dr = 0; dr < 4; dr++) {
                let nw = move(gr, x, y, dr);
                if (nw.c == '#' || (isDoor(nw.c) && !keys.has(keyForDoor(nw.c)))) continue;
                q.push({x: nw.x, y: nw.y, dist:dist+1});
            }
        }
    }
    return dst;
}

class State {
    constructor(pos, keys) {
        this.pos = pos;
        this.keys = keys;
    }
    toString() {
        return this.pos[0] + "," + this.pos[1] + "," + this.keys.n;
    }
    static fromString(s) {
        let [x, y, k] = s.split(",").map(f => parseInt(f));
        return new State([x, y], new KeySet(k));
    }
}

function partOne(inStr) {
    const [gr, loc, nKeys] = readData(inStr);
    let pathLengths = new Map();
    const state0 = new State(loc['@'], new KeySet(0));
    pathLengths.set(state0.toString(), 0);
    for (let depth = 0; depth < nKeys; depth++) {
        console.log("GEN:", depth, "SIZE:", pathLengths.size);
        const oldPathLengths = pathLengths;
        pathLengths = new Map();
        for (const [stateString,totalDist] of oldPathLengths) {
            const state = State.fromString(stateString);
            const kds = keyDistances(gr, state.pos[0], state.pos[1], state.keys);
            for (const [k, dist] of Object.entries(kds)) {
                const newTotalDist = totalDist + dist;
                const newPos = loc[k];
                const newKeys = (new KeySet(state.keys.n)).add(k);
                const newState = new State(newPos, newKeys);
                const newStateString = newState.toString();
                if (!(pathLengths.has(newStateString)) || newTotalDist < pathLengths.get(newStateString)) {
                    pathLengths.set(newStateString, newTotalDist);
                }
            }
        }
    }
    return Math.min(...pathLengths.values());
}

function runTests() {
    assert(isKey('b') && !isKey('Z'));
    assert(!isDoor('b') && isDoor('Z'));
    assert(keyForDoor('A') == 'a');

    const test1 = "###\n#.a\n.Z.";
    let [gr,loc,nKeys] = readData(test1);
    assert.deepEqual(gr, ["###", "#.a", ".Z."]);
    assert.deepEqual(loc, { a: [ 2, 1 ], Z: [ 1, 2 ] });
    assert(nKeys == 1);

    //assert.deepEqual(move(gr, 1,1, 1), [2, 1, 'a']);
    //assert.deepEqual(move(gr, 2,2, 2), [2, 3, '#']);

    let ks = new KeySet(0).add('c');
    console.log(ks);
    assert(ks.has('c') && !ks.has('a'));

    let st = new State([1,2], new KeySet(0));
    st.keys.add('c');
    assert(st.toString() == "1,2,4");
    let st2 = State.fromString(st.toString());
    console.log(st2);
    assert(st2.pos[0] == 1 && st2.pos[1] == 2 && st2.keys.n == 4);

    const test2 = `########################
    #f.D.E.e.C.b.A.@.a.B.c.#
    ######################.#
    #d.....................#
    ########################`;

    [gr,loc,nKeys] = readData(test2);
    let dst = keyDistances(gr, loc['@'][0], loc['@'][1], new KeySet().add('a').add('c'));
    assert.deepEqual(dst, {b: 4, e: 8});

    let ans2 = partOne(test2);
    assert(ans2 == 86);

    const test3 = `########################
    #...............b.C.D.f#
    #.######################
    #.....@.a.B.c.d.A.e.F.g#
    ########################`;
    let ans3 = partOne(test3);
    assert(ans3 == 132);

}

runTests();

let inputStr = fs.readFileSync('input.txt', 'utf8');
let ans1 = partOne(inputStr);
console.log("PART1:", ans1);
